src/bundle-manifest.ts

```diff
@@ -0,0 +1,150 @@
+ */
+export type BundleStrategy = (bundles: Bundle[]) => Bundle[];
+
+/**
+ * A mapping of entrypoints to their full set of transitive dependencies,
+ * such that a dependency graph `a->c, c->d, d->e, b->d, b->f` would be
+ * represented `{a:[a,c,d,e], b:[b,d,e,f]}`.  Please note that there is an
+ * explicit identity dependency (`a` depends on `a`, `b` depends on `b`).
+ */
+export type TransitiveDependenciesMap = Map<UrlString, Set<UrlString>>;
+
+/**
+ * Defining this URL type to make it clear which strings represent URLs.
+ */
+export type UrlString = string;
+
+/**
+ * A bundle is a grouping of files which serve the need of one or more
+ * entrypoint files.
+ */
+export class Bundle {
+  public entrypoints: Set<UrlString>;
+  public files: Set<UrlString>;
+  constructor(entrypoints?: Set<UrlString>, files?: Set<UrlString>) {
+    this.entrypoints = entrypoints || new Set<UrlString>();
+    this.files = files || new Set<UrlString>();
+  }
+}
+
+/**
+ * Given an index of files and their dependencies, produce an array of bundles,
+ * where a bundle is defined for each set of dependencies.
+ *
+ * For example, a dependency index representing the graph:
+ *   `a->b, b->c, b->d, e->c, e->f`
+ *
+ * Would produce an array of three bundles:
+ *   `[a]->[a,b,d], [e]->[e,f], [a,e]->[c]`
+ */
+export function generateBundles(depsIndex: TransitiveDependenciesMap):
+    Bundle[] {
+  const bundles: Bundle[] = [];
+  const invertedIndex = invertMultimap(depsIndex);
+
+  for (let entry of invertedIndex.entries()) {
+    let dep: UrlString = entry[0], entrypoints: Set<UrlString> = entry[1];
+    const entrypointsArray = Array.from(entrypoints);
+    let bundle = bundles.find((bundle) => {
+      return bundle.entrypoints.size === entrypoints.size &&
+          entrypointsArray.every((e) => bundle.entrypoints.has(e));
+    });
+    if (!bundle) {
+      bundle = new Bundle(entrypoints);
+      bundles.push(bundle);
+    }
+    bundle.files.add(dep);
+  }
+  return bundles;
+}
+
+/**
+ * Generates a strategy function to merge all bundles where the dependencies
+ * for a bundle are shared by at least 2 entrypoints (default; set
+ * `minEntrypoints` to change threshold).
+ *
+ * This function will convert an array of 4 bundles:
+ *   `[a]->[a,b], [a,c]->[d], [c]->[c,e], [f,g]->[f,g,h]`
+ *
+ * Into the following 3 bundles, including a single bundle for all of the
+ * dependencies which are shared by at least 2 entrypoints:
+ *   `[a]->[a,b], [c]->[c,e], [a,c,f,g]->[d,f,g,h]`
+ */
+export function generateSharedDepsMergeStrategy(minEntrypoints: number):
+    BundleStrategy {
+  return (bundles: Bundle[]): Bundle[] => {
+    const newBundles: Bundle[] = [];
+    const sharedBundles: Bundle[] = [];
+    for (let bundle of bundles) {
+      if (bundle.entrypoints.size >= minEntrypoints) {
+        sharedBundles.push(bundle);
+      } else {
+        newBundles.push(
+            new Bundle(new Set(bundle.entrypoints), new Set(bundle.files)));
```

Comment:
do you need to create a new Bundle? If `mergeBundles` doesn't mutate, but always returns a new Bundle, then it should be ok to just push `bundle`